File: backend/workflow/pipeline.go
package workflow

import (
	"context"
	"fmt"

	"github.com/dop251/goja"
	"github.com/enough/enough/backend/core"
)

type queuedJob struct {
	Key     string
	Phase   string
	Options AgentOptions
}

func (r *Runtime) runPipeline(ctx context.Context, vm *goja.Runtime, input any, stages []goja.Callable) (PipelineResult, error) {
	out := PipelineResult{
		Input: input, Results: map[string]AgentResult{},
	}
	var previous []AgentResult
	for index, stage := range stages {
		if err := ctx.Err(); err != nil {
			return out, err
		}
		phase := r.phaseName(index)
		r.setPhase(phase, index)
		stageCtx := map[string]any{
			"input":           input,
			"previousResults": previous,
			"results":         out.Results,
			"stageIndex":      index,
			"phase":           phase,
		}
		value, err := stage(goja.Undefined(), vm.ToValue(jsonShape(stageCtx)))
		if err != nil {
			return out, fmt.Errorf("pipeline stage %s: %w", phase, err)
		}
		value, err = unwrapPromise(value)
		if err != nil {
			return out, fmt.Errorf("pipeline stage %s: %w", phase, err)
		}
		jobs, err := normalizeJobs(vm, value, phase)
		if err != nil {
			return out, fmt.Errorf("pipeline stage %s: %w", phase, err)
		}
		results, err := r.runPool(ctx, phase, jobs)
		if err != nil {
			return out, err
		}
		previous = results
		out.Stages = append(out.Stages, StageResult{Name: phase, Results: results})
		for _, result := range results {
			out.Results[result.Key] = result
		}
	}
	return out, nil
}

func normalizeJobs(vm *goja.Runtime, value goja.Value, phase string) ([]queuedJob, error) {
	if value == nil || goja.IsUndefined(value) || goja.IsNull(value) {
		return nil, nil
	}
	var raw []map[string]any
	if err := exportJSONValue(value, &raw); err != nil {
		return nil, fmt.Errorf("stage must return an array of subjobs: %w", err)
	}
	jobs := make([]queuedJob, 0, len(raw))
	for index, item := range raw {
		optionsMap := item
		if nested, ok := item["options"].(map[string]any); ok {
			optionsMap = nested
			if _, exists := optionsMap["key"]; !exists {
				optionsMap["key"] = item["key"]
			}
		}
		var opts AgentOptions
		dataValue := vm.ToValue(optionsMap)
		if err := exportJSONValue(dataValue, &opts); err != nil {
			return nil, fmt.Errorf("subjob %d: %w", index, err)
		}
		if opts.Role == "" {
			opts.Role = phase
		}
		if opts.Key == "" {
			opts.Key = fmt.Sprintf("%s:%d", phase, index+1)
		}
		if opts.Prompt == "" {
			return nil, fmt.Errorf("subjob %s has no prompt", opts.Key)
		}
		jobs = append(jobs, queuedJob{Key: opts.Key, Phase: phase, Options: opts})
	}
	return jobs, nil
}

func (r *Runtime) phaseName(index int) string {
	r.mu.RLock()
	defer r.mu.RUnlock()
	if index >= 0 && index < len(r.meta.Phases) && r.meta.Phases[index] != "" {
		return r.meta.Phases[index]
	}
	return fmt.Sprintf("stage-%d", index+1)
}

func (r *Runtime) setPhase(phase string, index int) {
	r.mu.Lock()
	r.snapshot.Phase = phase
	if r.state != nil {
		r.state.Phase = phase
		r.state.StageIndex = index
	}
	r.ensurePhaseLocked(phase)
	r.mu.Unlock()
	r.persist()
	r.emitRun(core.EventWorkflowPhase, phase)
}

File: backend/workflow/pipeline.ts
// PORT: backend/workflow/pipeline.go

import { Effect } from "effect";
import { type PipelineResult, type StageResult, type AgentResult, type AgentOptions, cloneJSON } from "./types";
import { Runtime } from "./runtime";
import { event_workflow_phase } from "../core/events";

export interface queuedJob {
  Key: string;
  Phase: string;
  Options: AgentOptions;
}

export function runPipeline(
  this: Runtime,
  ctx: AbortSignal,
  input: any,
  stages: ((ctx: any) => Promise<any> | any)[]
): Effect.Effect<PipelineResult, Error> {
  const self = this;
  return Effect.gen(function* () {
    const out: PipelineResult = {
      input,
      stages: [],
      results: {},
    };

    let previous: AgentResult[] = [];

    for (let index = 0; index < stages.length; index++) {
      if (ctx.aborted) {
        return yield* Effect.fail(new Error("interrupted"));
      }
      const stage = stages[index];
      const phase = self.phaseName(index);
      self.setPhase(phase, index);

      const stageCtx = {
        input: input,
        previousResults: previous,
        results: out.results,
        stageIndex: index,
        phase: phase,
      };

      let value: any;
      try {
        const res = stage(cloneJSON(stageCtx));
        if (res && typeof res.then === "function") {
          value = yield* Effect.promise(() => res);
        } else {
          value = res;
        }
      } catch (err: any) {
        return yield* Effect.fail(new Error(`pipeline stage ${phase}: ${err.message || String(err)}`));
      }

      let jobs: queuedJob[];
      try {
        jobs = normalizeJobs(value, phase);
      } catch (err: any) {
        return yield* Effect.fail(new Error(`pipeline stage ${phase}: ${err.message || String(err)}`));
      }

      const results = yield* self.runPool(ctx, phase, jobs);
      previous = results;
      out.stages.push({ name: phase, results });
      for (const res of results) {
        if (res.key) {
          out.results[res.key] = res;
        }
      }
    }

    return out;
  });
}

export function normalizeJobs(value: any, phase: string): queuedJob[] {
  if (value === undefined || value === null) {
    return [];
  }
  if (!Array.isArray(value)) {
    throw new Error("stage must return an array of subjobs");
  }

  const jobs: queuedJob[] = [];
  for (let index = 0; index < value.length; index++) {
    const item = value[index];
    if (typeof item !== "object" || item === null) {
      throw new Error(`subjob ${index} is not an object`);
    }

    let optionsMap = { ...item };
    if (item.options && typeof item.options === "object") {
      optionsMap = { ...item.options };
      if (optionsMap.key === undefined) {
        optionsMap.key = item.key;
      }
    }

    const opts: AgentOptions = {
      key: optionsMap.key,
      role: optionsMap.role ?? "",
      prompt: optionsMap.prompt ?? "",
      systemPrompt: optionsMap.systemPrompt,
      tools: optionsMap.tools,
      model: optionsMap.model,
      responseSchema: optionsMap.responseSchema,
      maxTurns: optionsMap.maxTurns,
      readonly: optionsMap.readonly,
    };

    if (opts.role === "") {
      opts.role = phase;
    }
    if (!opts.key) {
      opts.key = `${phase}:${index + 1}`;
    }
    if (opts.prompt === "") {
      throw new Error(`subjob ${opts.key} has no prompt`);
    }

    jobs.push({ Key: opts.key, Phase: phase, Options: opts });
  }

  return jobs;
}

export function phaseName(this: Runtime, index: number): string {
  if (index >= 0 && this.meta.phases && index < this.meta.phases.length && this.meta.phases[index] !== "") {
    return this.meta.phases[index];
  }
  return `stage-${index + 1}`;
}

export function setPhase(this: Runtime, phase: string, index: number): void {
  this.snapshot.phase = phase;
  if (this.state) {
    this.state.phase = phase;
    this.state.stageIndex = index;
  }
  this.ensurePhaseLocked(phase);
  this.persist();
  this.emitRun(event_workflow_phase, phase);
}

// Attach prototype methods
const proto = Runtime.prototype as any;
proto.runPipeline = runPipeline;
proto.phaseName = phaseName;
proto.setPhase = setPhase;

/*
PORT STATUS
source path: backend/workflow/pipeline.go
source lines: 117
source lines: 111
confidence: high
status: phase_b_compile
*/
